File: java/revenj-servlet/src/main/java/org/revenj/server/ServerCommand.java
package org.revenj.server;

import org.revenj.serialization.Serialization;
import org.revenj.patterns.ServiceLocator;

import java.security.Principal;

public interface ServerCommand {
	<TInput, TOutput> CommandResult<TOutput> execute(
			ServiceLocator locator,
			Serialization<TInput> input,
			Serialization<TOutput> output,
			TInput data,
			Principal principal);
}

File: scala/revenj-akka/src/main/scala/net/revenj/server/ServerCommand.scala
package net.revenj.server

import net.revenj.patterns.ServiceLocator
import net.revenj.serialization.Serialization

import scala.concurrent.Future

trait ServerCommand {
  def execute[TInput, TOutput](
    locator: ServiceLocator,
    input: Serialization[TInput],
    output: Serialization[TOutput],
    data: TInput): Future[CommandResult[TOutput]]
}

trait ReadOnlyServerCommand extends ServerCommand
